src/util.c

```diff
@@ -1268,15 +1268,22 @@ int rename_noreplace(int olddirfd, const char *oldpath, int newdirfd, const char
                 return -errno;
 
         /* Let's try linkat(). This will of course failure for non-files, but that's fine. */
-        if (linkat(olddirfd, oldpath, newdirfd, newpath, 0) < 0)
-                return -errno;
+        if (linkat(olddirfd, oldpath, newdirfd, newpath, 0) >= 0) {
+                if (unlinkat(olddirfd, oldpath, 0) >= 0)
+                        return 0;
 
-        if (unlinkat(olddirfd, oldpath, 0) < 0) {
                 r = -errno;
                 (void) unlinkat(newdirfd, newpath, 0);
                 return r;
+        } else if (!ERRNO_IS_UNSUPPORTED(errno)) {
+                return -errno;
         }
 
+        /* if renameat2 and linkat aren't supported, fallback to renameat,
+         * even if it mean maybe overwriting an existing file */
+        if (renameat(olddirfd, oldpath, newdirfd, newpath) < 0)
+                return -errno;
```

Comment:
hmm, can you add a check with faccessat() first, i.e. so that we won't replace things, which is what this call is about after all. It's not going to be atomic in that case, but it's still better to have the right behaviour though non-atomic than the wrong behaviour.